tools/api_proto_plugin/type_context.py

```diff
@@ -60,22 +60,34 @@ def LocationPathLookup(self, path):
 
   # TODO(htuch): consider integrating comment lookup with overall
   # FileDescriptorProto, perhaps via two passes.
-  def LeadingCommentPathLookup(self, path):
+  def LeadingCommentPathLookup(self, path, annotation_xforms=None):
     """Lookup leading comment by path in SourceCodeInfo.
 
     Args:
       path: a list of path indexes as per
         https://github.com/google/protobuf/blob/a08b03d4c00a5793b88b494f672513f6ad46a681/src/google/protobuf/descriptor.proto#L717.
+      annotation_xforms: a dict of transformers to process annotation extracted
 
     Returns:
       Comment object.
     """
     location = self.LocationPathLookup(path)
-    if location is not None:
-      return Comment(
-          location.leading_comments,
-          annotations.ExtractAnnotations(location.leading_comments, self.file_level_annotations))
-    return Comment('', {})
+    if location is None:
+      return Comment('', '', {})
+    raw = xformed = location.leading_comments
+    annotation_map = annotations.ExtractAnnotations(raw, self.file_level_annotations)
+    if annotation_xforms:
+      for annotation, xform in sorted(annotation_xforms.items()):
+        original_annotation = annotation_map.get(annotation)
+        xformed_annotation = xform(original_annotation)
+        if original_annotation is not None:
+          xformed = annotations.WithoutAnnotation(xformed, annotation)
+          del annotation_map[annotation]
+        if xformed_annotation is not None:
+          xformed += annotations.FormatAnnotation(annotation, xformed_annotation)
```

Comment:
Can you add some comments here? The logic is a bit confusing, in particular since there is a mutation of `annotation_map` via `del` (why?).

Reply:
Just split the mutation to two part, del and put. Otherwise, more mix check needed.
e.g. if the transformer want to remove the annotation by returning `None`, then just del it.